tuning-fork/boilerplate
# 467/add cypress testing

## 🤺 Summary
This PR covers the implementation of Cypress E2E testing for the Boilerplate React UI:

- Installed Cypress

- Created Cypress test coverage for page-level features & CRUD operations:
- Categories
- Funding Orgs
- Boilerplates
- Grants, including:
- Copy Grant
- Grant Sections
- Grant Overview (Drag&Drop and Preview features)
- Paste Boilerplate

There is also an associated API branch that covers the creation of seed data for the tests:
[feature/cypress/474-add-cypress-db-seeds-v2](https://github.com/tuning-fork/boilerplate_api/tree/feature/cypress/474-add-cypress-db-seeds)

### Steps to Test:
- Make sure that you are checked out to the API branch
- Run rails db:reset to set/reset the test data

- Install Cypress:
npm install cypress

- Run Cypress:
npm run cypress:open
(Make sure both rails and react servers are running :))

- Run all test files in the e2e directory
npx cypress run -s cypress/e2e/boilerplate-e2e

- To run specific files, append the file name
npx cypress run -s cypress/e2e/boilerplate-e2e/<test file name>

Cypress runs in a separate browser. It should open automatically, and you should see a browser interface. The landing page offers you a choice between E2E and Component Testing. 

![image](https://github.com/tuning-fork/boilerplate/assets/58121322/bfeb2e0c-686d-4f6f-89ca-3b0dc87cdb69)

Once you choose E2E testing, it will ask you to select a browser (we're assuming Chrome).

![image](https://github.com/tuning-fork/boilerplate/assets/58121322/b989efb8-e47a-4dd8-8558-7bbbd6f1077a)

Once you click through to Chrome, you should see an index menu that looks a little bit like the menu for Storybook.js. You will see an index of all of the test files; this index is automatically generated and reflects the Cypress test directories in the codebase. You'll also see directories that contain out of the box examples and test recipes, as well as some scratchpads for DnD; we're keeping those for now:

![image](https://github.com/tuning-fork/boilerplate/assets/58121322/666422a8-55b9-4ffa-ad34-3a8906809a30)

You can run tests by clicking on the individual test files. 

Once you start running tests, the UI will show a 'speedrun' of the test operations, as in the app:

![image](https://github.com/tuning-fork/boilerplate/assets/58121322/a3abeda5-7051-4da4-ac19-119d5beb70d4)

![image](https://github.com/tuning-fork/boilerplate/assets/58121322/f2495c30-e432-4515-b0d1-03c89502fb98)

The Cypress browser interface also includes a console on the lefthand side that shows a line by line readout of the test run. 

You can step back through each line; the screen should show what is happening during that step, with any selected elements highlighted, etc. You can also open up the line-by-line items in the console to see additional info, like data, selectors, etc. 

Test steps that pass will show up in the console as green; if the test hits an error, it will halt and you will see a red console printout. 

On the righthand side of the screen, there is also a more conventional devtools console; you can use this to run console.logs, etc.
